### PR TITLE
feat: create replicas when nodes become available

### LINKED ISSUE #306: bug: Replica mount should reset lost replicas once nodes are available [ETA 01/28/2022]
**What happened**:
3 node AKS cluster (both non-zonal and zonal). Statefulset with pvc using storageclass supporting replica mounts. replicas mounts set to 3 == the number of cluster nodes. When one node is lost or cordoned off, the number of replicas for the volume goes down to 2. When the node comes back up the number of replica attachments does not go back up to 3. 

**What you expected to happen**:
When the node comes back up the number of replica attachments goes back up to 3. 

**How to reproduce it**:

**Anything else we need to know?**:

**Environment**:
- CSI Driver version:
- Kubernetes version (use `kubectl version`):
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by kassarl**
**What type of PR is this?**

/kind feature
**What this PR does / why we need it**:
Creates a controller to watch for new node availability. Upon a node being created or uncordoned, uses a queue structure to create new replica attachments on that node
**Which issue(s) this PR fixes**:

Fixes #306 

**Requirements**:
- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

File: pkg/controller/common.go [hunk 1]
```diff
@@ -305,6 +308,64 @@ func (c *SharedState) deleteOperationQueue(volumeName string) {
 	lockable.Unlock()
 }
 
+func (c *SharedState) createReplicaRequestsQueue() {
+	heap.Init(&c.priorityReplicaRequestsQueue.queue)
+	heap.Init(&c.failedReplicaRequestsQueue.queue)
+}
+
+func (c *SharedState) generateReplicaAttachmentRoutines(ctx context.Context, azr azReconciler, namespace string) error {
```
Comment (edreed) [inline pkg/controller/common.go:316]: While technically correct that `generateReplicaAttachmentRoutines` creates `goroutine`, this is an implementation detail of processing the pending replicas. Consider `processPendingReplicaQueues` instead.

File: pkg/controller/common.go [hunk 2]
```diff
@@ -305,6 +308,64 @@ func (c *SharedState) deleteOperationQueue(volumeName string) {
...
+	klog.V(5).Infof("Begin dequeue from replica attachment queue for nodes")
+	nodes := &v1.NodeList{}
+	err := azr.getClient().List(ctx, nodes)
+	if err != nil {
+		klog.V(5).Infof("Unable to get nodes")
+		return err
+	}
+
+	for _, node := range nodes.Items {
```
Comment (edreed) [inline pkg/controller/common.go:325]: Scheduling replicas to the correct nodes needs to factor in pod taints/toleratation and affinity/anti-affinity. This logic is currently implemented by [`func (r *ReconcileReplica) manageReplicas(ctx context.Context, volumeName string) error`](https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/9eedc76217f699d055040f587a7e884f1141c646/pkg/controller/replica.go#L185). Ideally, process the queues and invoke this function rather than replicate the logic here. I wonder if we should factor the replica management code into `common.go` to make it easier to consume here? @sunpa93 ?

File: pkg/controller/common.go [hunk 3]
```diff
@@ -1259,3 +1320,64 @@ func containsString(key string, items []string) bool {
...
+
+type VolumeReplicaRequestsPriorityQueue struct {
+	mut   *sync.Mutex
+	queue PriorityQueue
+}
+
+func (vq VolumeReplicaRequestsPriorityQueue) Push(replicaRequest ReplicaRequest) {
+	vq.mut.Lock()
+	defer vq.mut.Unlock()
+	vq.queue.Push(replicaRequest)
```
Comment (edreed) [inline pkg/controller/common.go:1339]: This needs to be done through [`heap.Push()`](https://pkg.go.dev/container/heap#Push) in order to maintain the correct order, i.e.:
```go
heap.Push(&vq.queue, replicaRequest)
```

File: pkg/controller/common.go [hunk 4]
```diff
@@ -1259,3 +1320,64 @@ func containsString(key string, items []string) bool {
...
+	return
+}
+
+func (vq VolumeReplicaRequestsPriorityQueue) Pop() *ReplicaRequest {
+	vq.mut.Lock()
+	defer vq.mut.Unlock()
+	return vq.queue.Pop().(*ReplicaRequest)
```
Comment (edreed) [inline pkg/controller/common.go:1346]:
```go
return heap.Pop(&vq.queue).(*ReplicaRequest)
```

File: pkg/controller/common.go [hunk 5]
```diff
@@ -1259,3 +1320,64 @@ func containsString(key string, items []string) bool {
...
+	priority      int               //The number of replicas that have yet to be created
+	index         int
+}
+
+type VolumeReplicaRequestsPriorityQueue struct {
+	mut   *sync.Mutex
+	queue PriorityQueue
+}
+
+func (vq VolumeReplicaRequestsPriorityQueue) Push(replicaRequest ReplicaRequest) {
```
Comment (edreed) [inline pkg/controller/common.go:1336]: Pass by reference (i.e. pointer) here.

File: pkg/controller/common.go (hunk 2 above)
Comment (sunpa93) [reply pkg/controller/common.go:325]: Ideally, I think the replica creation processes in 1) pod controller 2) replica controller 3) node availability controller should be all combined into a single function in `common.go` by leveraging the `priorityReplicaRequestsQueue` instead of maintaining 3 separate replica creation work flows.

File: pkg/controller/common.go [hunk 6]
```diff
@@ -1259,3 +1320,64 @@ func containsString(key string, items []string) bool {
...
+	return
+}
+
+func (vq VolumeReplicaRequestsPriorityQueue) Pop() *ReplicaRequest {
+	vq.mut.Lock()
+	defer vq.mut.Unlock()
+	return vq.queue.Pop().(*ReplicaRequest)
+}
+
+type PriorityQueue []*ReplicaRequest
```
Comment (edreed) [inline pkg/controller/common.go:1349]: It makes sense to create a generic priority queue implementation in `common.go` or even `pkg/util` rather that tie it specifically to `ReplicaRequest`. You can also omit the self-index tracking and update mechanism that the example has since we do not need it.

```go
type LessFunc func(left, right interface{}) bool

type itemsArray struct {
	lessFn LessFunc
	items  []interface{}
}

func (i *itemsArray) Len() int {
	return len(i.items)
}

func (i *itemsArray) Less(left, right int) bool {
	return i.lessFn(i.items[left], i.items[right])
}

func (i *itemsArray) Swap(x, y int) {
	i.items[x], i.items[y] = i.items[y], i.items[x]
}

func (i *itemsArray) Push(value interface{}) {
	i.items = append(i.items, value)
}

func (i *itemsArray) Pop() interface{} {
	old := i.items
	n := len(old)
	value := old[n-1]
	old[n-1] = nil
	i.items = old[0 : n-1]

	return value
}

type PriorityQueue struct {
	items itemsArray
}

func NewPriorityQueue(lessFn LessFunc) *PriorityQueue {
	pq := &PriorityQueue{items: itemsArray{lessFn: lessFn}}

	heap.Init(&pq.items)

	return pq
}

func (pq *PriorityQueue) Push(value interface{}) {
	heap.Push(&pq.items, value)
}

func (pq *PriorityQueue) Pop() interface{} {
	return heap.Pop(&pq.items)
}
```

With this implementation, you create a priority queue for `ReplicaRequest` like this:

```go
queue := NewPriorityQueue(func (left, right interface{}) bool {
  return left.(*ReplicaRequest).priority > right.(*ReplicaRequest).priority
})
```

File: pkg/controller/common.go [hunk 7]
```diff
@@ -305,6 +308,64 @@ func (c *SharedState) deleteOperationQueue(volumeName string) {
 	lockable.Unlock()
 }
 
+func (c *SharedState) createReplicaRequestsQueue() {
+	heap.Init(&c.priorityReplicaRequestsQueue.queue)
+	heap.Init(&c.failedReplicaRequestsQueue.queue)
+}
+
+func (c *SharedState) generateReplicaAttachmentRoutines(ctx context.Context, azr azReconciler, namespace string) error {
+	klog.V(5).Infof("Begin dequeue from replica attachment queue for nodes")
```
Comment (sunpa93) [inline pkg/controller/common.go:317]: I think we should leverage using semaphores to prevent multiple instances of `generateReplicaAttachmentRoutines` running concurrently.

File: pkg/controller/node_availability.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+		Log:                     logger,
+	})
+
+	if err != nil {
+		klog.Errorf("Failed to create node availability controller. Error: %v", err)
+		return nil, err
+	}
+
+	//Predicate
+	p := predicate.Funcs{
```
Comment (edreed) [inline pkg/controller/node_availability.go:89]: We're only interested in processing the replica priority queues on node creation and changes from unschedulable to schedulable. `CreateFunc` should unconditionally return `true` (as it does currently), `UpdateFunc` should only return true if the previous `Node.Unschedulable` state was `false` and current is now `true`, and the remaining funcs should unconditionally return `false`.

**Comment by sunpa93** [comment]
We should also add `failedReplicaAttachments` back to `priorityReplicaRequeustsQueue` after a volume is garbage collected in replica.go
https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/9eedc76217f699d055040f587a7e884f1141c646/pkg/controller/replica.go#L155-L170

File: pkg/controller/common.go (hunk 7 above)
Comment (edreed) [reply pkg/controller/common.go:317]: See https://pkg.go.dev/golang.org/x/sync/semaphore.

File: pkg/util/util.go [hunk 9]
```diff
@@ -277,3 +277,37 @@ func ErrorFromAzError(azError *v1alpha1.AzError) error {
...
+}
+
+func (i *ItemsArray) Pop() interface{} {
+	old := i.Items
+	n := len(old)
+	value := old[n-1]
+	old[n-1] = nil
+	i.Items = old[0 : n-1]
+	return value
+}
```
Comment (edreed) [inline pkg/util/util.go:313]: Missing the actual priority queue implementation:

```go
type PriorityQueue struct {
	items itemsArray
}

func NewPriorityQueue(lessFn LessFunc) *PriorityQueue {
	pq := &PriorityQueue{items: itemsArray{lessFn: lessFn}}

	heap.Init(&pq.items)

	return pq
}

func (pq *PriorityQueue) Push(value interface{}) {
	heap.Push(&pq.items, value)
}

func (pq *PriorityQueue) Pop() interface{} {
	return heap.Pop(&pq.items)
}
```

File: pkg/controller/common.go [hunk 10]
```diff
@@ -1259,3 +1271,170 @@ func containsString(key string, items []string) bool {
...
+	vq.queue.Push(replicaRequest)
+}
+
+func (vq VolumeReplicaRequestsPriorityQueue) Pop() *ReplicaRequest {
+	vq.mut.Lock()
+	defer vq.mut.Unlock()
+	return vq.queue.Pop().(*ReplicaRequest)
+}
+
+type PriorityQueue struct {
```
Comment (edreed) [inline pkg/controller/common.go:1297]: The general priority queue implementation should be moved to `util.go` as well.

File: pkg/controller/node_availability.go [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+
+	n := &corev1.Node{}
+	err := r.client.Get(ctx, request.NamespacedName, n)
+
+	if errors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	}
+
+	if err == nil {
+		//Node is schedulable, proceed to attempt creation of replica attachment
```
Comment (edreed) [inline pkg/controller/node_availability.go:55]: You still need to check `node.Unschedulable` here even though you're also checking in the predicates because the node may have been created with `Unschedulable` set to `true` and we're passing creation events through unconditionally.

File: pkg/controller/node_availability.go [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+		r.controllerSharedState.priorityReplicaRequestsQueue.mut.Lock()
+		for r.controllerSharedState.priorityReplicaRequestsQueue.queue.Len() > 0 {
+			replicaRequest := r.controllerSharedState.priorityReplicaRequestsQueue.queue.Pop().(*ReplicaRequest)
```
Comment (edreed) [inline pkg/controller/node_availability.go:58]: Use `r.controllerSharedState.priorityReplicaRequestQueue.Pop()` to properly synchronize with other goroutines potentially modifying the queue.

File: pkg/controller/node_availability.go [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+	n := &corev1.Node{}
+	err := r.client.Get(ctx, request.NamespacedName, n)
+
+	if errors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	}
+
+	if err == nil {
+		//Node is schedulable, proceed to attempt creation of replica attachment
+		r.controllerSharedState.priorityReplicaRequestsQueue.mut.Lock()
```
Comment (edreed) [inline pkg/controller/node_availability.go:56]: Using a mutex to prevent re-entry into the queue processing loop is overkill and can lead to blocked goroutines that will end up doing nothing when they acquire the lock (i.e. first goruotine processes and clears the queue then releases the mutex. Unblocked goroutines now have nothing to do.)

You can use a the [atomic](https://pkg.go.dev/sync/atomic) package and a simple `int` flag in this case, e.g.

```go
// Set the flag to 1 and process the queue if it was previously 0
if atomic.SwapInt32(&r.controllerSharedState.processingReplicaRequestQueue, 1) == 0 {
  // Process the priority queue here

  // Set the flag back to 0 to allow other qoroutines to enter the processing loop
  atomic.StoreInt32(&r.controllerSharedState.processingReplicaRequestQueue, 0)
}
```

File: pkg/controller/common.go [hunk 14]
```diff
@@ -1259,3 +1278,152 @@ func containsString(key string, items []string) bool {
...
+type ReplicaRequest struct {
+	VolumeName string
+	Priority   int //The number of replicas that have yet to be created
+}
+type VolumeReplicaRequestsPriorityQueue struct {
+	mut   *sync.Mutex
+	queue *cache.Heap
+}
+
+func (vq VolumeReplicaRequestsPriorityQueue) Push(replicaRequest *ReplicaRequest) {
```
Comment (edreed) [inline pkg/controller/common.go:1291]: The receiver parameter needs to be a pointer here, i.e. `vq *VolumeReplicaRequestsPriorityQueue`. As implemented, this function updates a local copy of the `VolumeReplicaRequestsPriorityQueue`, so it has no effect on the target object.

File: pkg/controller/common.go [hunk 15]
```diff
@@ -305,6 +313,17 @@ func (c *SharedState) deleteOperationQueue(volumeName string) {
 	lockable.Unlock()
 }
 
+func (c *SharedState) createReplicaRequestsQueue() {
+	c.priorityReplicaRequestsQueue.queue = cache.NewHeap(
```
Comment (edreed) [inline pkg/controller/common.go:317]: Nice find! Per the [Heap](https://pkg.go.dev/k8s.io/client-go@v0.23.1/tools/cache#Heap) documentation, it is thread-safe. Do we need the mutex at all?

File: pkg/controller/common_test.go [hunk 16]
```diff
@@ -324,7 +342,7 @@ func createPod(podNamespace, podName string, pvcs []string) v1.Pod {
 
 func initState(objs ...runtime.Object) (c *SharedState) {
 	c = NewSharedState(consts.DefaultDriverName)
-
+	c.createReplicaRequestsQueue()
```
Comment (edreed) [inline pkg/controller/common_test.go:345]: This is already done by `NewSharedState` and isn't needed here.

File: pkg/controller/node_availability.go [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+	if errors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	}
+
+	if err == nil {
+		if !n.Spec.Unschedulable {
+			//Node is schedulable, proceed to attempt creation of replica attachment
+			if atomic.SwapInt32(&r.controllerSharedState.processingReplicaRequestQueue, 1) == 0 {
+				for len(r.controllerSharedState.priorityReplicaRequestsQueue.queue.List()) > 0 {
+					replicaRequest := r.controllerSharedState.priorityReplicaRequestsQueue.Pop()
```
Comment (edreed) [inline pkg/controller/node_availability.go:62]: It's unfortunate that `cache.Heap.Pop` is blocking thus forcing you to `List` in order to see if the queue is empty. This can be a relatively expensive operation. Consider tracking the count of items in the queue separately in `VolumeReplicaRequestsPriorityQueue`. You can use [`atomic.AddInt32`](https://pkg.go.dev/sync/atomic#AddInt32) to update the count in the `Push` and `Pop` methods.

File: test/e2e/dynamic_provisioning_test.go [hunk 18]
```diff
@@ -1286,4 +1286,60 @@ func (t *dynamicProvisioningTestSuite) defineTests(isMultiZone bool, schedulerNa
 		}
 		test.Run(cs, ns, schedulerName)
 	})
+	ginkgo.It("Laila Should test an increase in replicas when scaling up", func() {
```
Comment (edreed) [inline test/e2e/dynamic_provisioning_test.go:1289]: Planning to commit yourself to the repository to run the tests? 😆

**Review by edreed — CHANGES_REQUESTED** [review]
I didn't see an obvious reason for the failure to create the primary `AzVolumeAttachment`. However, the priority queue implementation doesn't work because you're using a receiver by value. See https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/pull/333/files#r779809946.

File: test/e2e/dynamic_provisioning_test.go (hunk 18 above)
Comment (kassarl) [reply test/e2e/dynamic_provisioning_test.go:1289]: Ope guess I'll have to leave my legacy elsewhere

File: pkg/controller/common.go (hunk 15 above)
Comment (kassarl) [reply pkg/controller/common.go:317]: Yep seems like we won't, I'll give it a try without the mutex 👍

File: pkg/controller/node_availability.go (hunk 17 above)
Comment (kassarl) [reply pkg/controller/node_availability.go:62]: I'm also wondering if, with my implementation, the operation in addToOperationQueue completes before this loop is finished, and the attachment fails and volume is added back to the queue, this would be an infinite loop? or, if it requeues the request and the priority is still higher than the other volumes, this would also never reach those lower ranked volumes. Any thoughts? I could store volumes already tried and check against that before attempting addToOperationQueue, not sure if that would be expensive

File: test/types/types.go [hunk 19]
```diff
@@ -402,3 +402,18 @@ func GetReplicaAttachments(persistentVolume *v1.PersistentVolume, client clients
 	}
 	return azVolumeAttachmentsReplica, nil
 }
+
+func GetSchedulableNodes(azDiskClient *azDiskClientSet.Clientset, client clientset.Interface) []*v1.Node {
+	nodes := nodeutil.ListAzDriverNodeNames(azDiskClient.DiskV1alpha1().AzDriverNodes(azureconstants.AzureDiskCrdNamespace))
+	var availableNodes []*v1.Node
+
+	for _, nodeName := range nodes {
+		node, err := client.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
```
Comment (landreasyan) [inline test/types/types.go:411]: Instead of issuing a `Get` request for each node, `List` all the nodes with `metav1.ListOptions{}` using a `FieldSelector` for the `Spec.Unschedulable` field.

File: pkg/controller/node_availability.go [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+				return reconcile.Result{Requeue: false}, nil
+			}
+		}
+	}
+	return reconcile.Result{Requeue: false}, err
+}
+
+func NewNodeAvailabilityController(mgr manager.Manager, azVolumeClient azClientSet.Interface, kubeClient kubernetes.Interface, namespace string, controllerSharedState *SharedState) (*ReconcileNodeAvailability, error) {
+	logger := mgr.GetLogger().WithValues("controller", "nodeavailability")
+	reconciler := ReconcileNodeAvailability{
```
Comment (landreasyan) [inline pkg/controller/node_availability.go:83]: Nodes are not namespaced objects. Do we need the namespace parameter here?

File: pkg/controller/node_availability.go (hunk 17 above)
Comment (edreed) [reply pkg/controller/node_availability.go:62]: Yeah, that's a good point and we'll need to defend against this scenario. I don't really like the idea of an exclusion list, but I don't have any ideas that still allow us to use `client-go/tools/cache/Heap`. If you go back to using our own priority queue implementation, we could implement a `func Drain() []*ReplicaRequest` function that would atomically empty and return all elements on the queue as an array in priory order. This would allow concurrent and requeue requests to be made w/o resulting in an infinite loop.

File: pkg/controller/pod.go [hunk 21]
```diff
@@ -115,17 +109,10 @@ func (r *ReconcilePod) createReplicas(ctx context.Context, podKey string) error
...
-				for _, node := range nodes {
-					if numCreated >= azVolume.Spec.MaxMountReplicaCount {
-						break
-					}
-					if err := createReplicaAzVolumeAttachment(ctx, r, azVolume.Status.Detail.ResponseObject.VolumeID, node, azVolume.Spec.Parameters); err != nil {
-						klog.Warningf("Error creating %d/%d replicas azvolumeattachment for pod %s and volume %s on node %s. Error: %v", azVolume.Spec.MaxMountReplicaCount, numCreated, podKey, volume, node, err)
-						return err
-					}
-					numCreated++
+				err = r.controllerSharedState.manageReplicas(ctx, azVolume.Spec.UnderlyingVolume, r, r.namespace)
```
Comment (edreed) [inline pkg/controller/pod.go:112]: Looking at the e2e test failures, it seems the reason no replicas are create is because `manageReplicas` can't find the `AzVolume` instance:

```log
2022-01-10T12:51:20.4438991Z I0110 11:55:39.870754       1 pod.go:87] Pod azuredisk-1386/azuredisk-volume-tester-8kpq4 has 1 volumes. Volumes: [pvc-0f90b76a-b9bf-4982-b581-893f04ea13a5]
2022-01-10T12:51:20.4441451Z I0110 11:55:39.870764       1 pod.go:75] Successfully created replicas for pod azuredisk-volume-tester-8kpq4. Reconciliation succeeded.
2022-01-10T12:51:20.4443075Z I0110 11:55:39.870813       1 common.go:1312] Aborting replica management... volume (pvc-0f90b76a-b9bf-4982-b581-893f04ea13a5) does not exist
2022-01-10T12:51:20.4445507Z I0110 11:55:39.870823       1 common.go:299] operation queue for volume (pvc-0f90b76a-b9bf-4982-b581-893f04ea13a5) fully iterated and completed.
```

Here `r.namespace` is passed to `managedReplicas` but it is not initialized by this reconciler because pods are not specific to the Mandalay driver's namespace. You'll need to pass the namespace here. It's not currently available from the shared state but should be. You can get it from the `driver.objectNamespace` field in `newDriverV2` when initializing the shared state

Comment (edreed) [reply pkg/controller/pod.go:112]: @kassarl I noticed that we were inconsistently using the `--driver-object-namespace` parameter which allows specifying the CRD namespace and a hardcoded constant, so I went and complete support for customizing the CRD namespace in PR #340. It includes adding the CRD namespace to the shared state.

File: pkg/controller/common.go [hunk 22]
```diff
@@ -1263,3 +1283,150 @@ func containsString(key string, items []string) bool {
...
+
+	nodes, err := c.getNodesForReplica(ctx, volumeName, pods, numReplica, azr)
+	if err != nil {
+		klog.Errorf("failed to get a list of nodes for replica attachment: %v", err)
+		return err
+	}
+
+	numCreated := 0
+	for _, node := range nodes {
+		if err := createReplicaAzVolumeAttachment(ctx, azr, volumeID, node, volumeContext); err != nil {
```
Comment (nearora-msft) [inline pkg/controller/common.go:1376]: This will handle the case where the CRI creation failed. Can there be cases when the CRI was created successfully but the actual attachment failed and are we adding the pending replicas to the queue for that case as well?

Comment (sunpa93) [reply pkg/controller/common.go:1376]: Currently, we handle failed attachments for replicas in replica controller by deleting the replica CRI and creating a replacement replica. I suppose we could think about integrating this process into the new queueing method.

File: pkg/controller/node_availability.go [hunk 23]
```diff
@@ -0,0 +1,155 @@
...
+	n := &corev1.Node{}
+	err := r.client.Get(ctx, request.NamespacedName, n)
+
+	if errors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	}
+
+	if err == nil {
+		if !n.Spec.Unschedulable {
+			for _, taint := range n.Spec.Taints {
```
Comment (edreed) [inline pkg/controller/node_availability.go:58]: Taints don't prevent every pod from being scheduled to this node, so we shouldn't exclude it here based on the presence of a no-schedule taint.

File: test/types/types.go [hunk 24]
```diff
@@ -402,3 +403,35 @@ func GetReplicaAttachments(persistentVolume *v1.PersistentVolume, client clients
 	}
 	return azVolumeAttachmentsReplica, nil
 }
+
+func GetSchedulableNodes(azDiskClient *azDiskClientSet.Clientset, client clientset.Interface) []*v1.Node {
```
Comment (edreed) [inline test/types/types.go:407]: Ideally, this function should take a `v1.Pod` and check for node taints with no matching toleration on the pod. You can use the `Toleration.ToleratesTaint()` method to check if a toleration satisfies the taint criteria.

File: pkg/controller/node_availability.go (hunk 23 above)
Comment (kassarl) [reply pkg/controller/node_availability.go:58]: Thats a good point, I will update this to just check for the node.kubernetes.io/unschedulable taint, as that should still allow us to requeue the request until the cache with this taint is updated to match what is reflected in n.Spec.Unschedulable

Comment (edreed) [reply pkg/controller/node_availability.go:58]: Good to know about the `node.kubernetes.io/unschedulable` taint. You'll need to update the update predicate on line 121 to return `true` when taints change then. (The `Unschedulable` property is updated by the user; presumably, the taint is added or removed by the Node reconciler in response. This means we might see the change in Unschedulable before the taint has been added/removed.)

Comment (kassarl) [reply pkg/controller/node_availability.go:58]: As another thought, should we also be using the node_availability controller to try to add replicas when other taints are updated other than the basic node.kubernetes.io/unschedulable node scheduling? For example, when any taint gets updated on the node, the node_availability controller could try to add replica attachments from the priority queue to see if any are newly tolerated

Comment (edreed) [reply pkg/controller/node_availability.go:58]: I don't believe it very likely that taints will be changed often, and when they do, it mostly to support new workloads that have yet to be scheduled. I don't believe we need to consider that scenario in this PR nor in the future.

File: pkg/controller/replica.go [hunk 25]
```diff
@@ -138,7 +136,7 @@ func (r *ReconcileReplica) Reconcile(ctx context.Context, request reconcile.Requ
 func (r *ReconcileReplica) triggerGarbageCollection(volumeName string) {
```
Comment (sunpa93) [inline pkg/controller/replica.go:136]: As replica `AzVolumeAttachments` for a volume gets garbage collected, disk attach capacity of nodes to which the replicas were previously attached gets freed. So, ideally, we should create failed replicas here as well.

File: pkg/controller/node_availability.go [hunk 26]
```diff
@@ -0,0 +1,155 @@
...
+				if taint.Effect == corev1.TaintEffectNoSchedule {
+					return reconcile.Result{Requeue: true}, nil
+				}
+			}
+
+			//Node is schedulable, proceed to attempt creation of replica attachment
+			if atomic.SwapInt32(&r.controllerSharedState.processingReplicaRequestQueue, 1) == 0 {
```
Comment (sunpa93) [inline pkg/controller/node_availability.go:65]: Maybe what comes below fits better as a method under `VolumeReplicaRequestsPriorityQueue` receiver (e.g. `func (vq *VolumeReplicaRequestsPriorityQueue) tryCreateFailedReplicas(...)`).  That way, this procedure can be used elsewhere (e.g. after garbage collection)

**Review by sunpa93 — CHANGES_REQUESTED** [review]

File: pkg/controller/node_availability.go [hunk 27]
```diff
@@ -0,0 +1,155 @@
...
+
+	if errors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	}
+
+	if err == nil {
+		if !n.Spec.Unschedulable {
+			for _, taint := range n.Spec.Taints {
+				if taint.Effect == corev1.TaintEffectNoSchedule {
+					return reconcile.Result{Requeue: true}, nil
```
[the remark this answers is not in the document]
Comment (sunpa93) [reply pkg/controller/node_availability.go:60]: Oh, nvm. This has already been discussed between Ed and you :)

File: pkg/controller/replica.go (hunk 25 above)
Comment (kassarl) [reply pkg/controller/replica.go:136]: Good point, I was planning on creating another PR to address creating failed replicas after garbage collection to keep this one in the scope of the node_availability controller

File: pkg/controller/node_availability.go (hunk 26 above)
Comment (kassarl) [reply pkg/controller/node_availability.go:65]: Good point, I've moved it now to the common.go

File: test/e2e/testsuites/dynamically_provisioned_pod_failover_with_replicas_tester.go [hunk 28]
```diff
@@ -75,11 +75,23 @@ func (t *PodFailoverWithReplicas) Run(client clientset.Interface, namespace *v1.
...
 	err := wait.Poll(15*time.Second, 10*time.Minute, func() (bool, error) {
+		failedReplicaAttachments = nil
+		allReplicasAttached = true
 		var err error
-		allReplicasAttached, failedReplicaAttachments, err = testtypes.VerifySuccessfulReplicaAzVolumeAttachments(t.Pod, t.AzDiskClient, t.StorageClassParameters, client, namespace)
+		var attached bool
+		var podFailedReplicaAttachments *v1alpha1.AzVolumeAttachmentList
+		for _, pod := range tDeployment.Pods {
+			attached, podFailedReplicaAttachments, err = testtypes.VerifySuccessfulReplicaAzVolumeAttachments(pod, t.AzDiskClient, t.StorageClassParameters, client, namespace)
+			allReplicasAttached = allReplicasAttached && attached
```
Comment (edreed) [inline test/e2e/testsuites/dynamically_provisioned_pod_failover_with_replicas_tester.go:89]: `allReplicasAttached` defaults to `false` on line 78 above, so it can never be set to `true` in this expression.

File: test/types/resource_types.go [hunk 29]
```diff
@@ -479,18 +479,18 @@ func (t *TestDeployment) Create() {
 	pods, err := podutil.GetPodsForDeployment(t.Client, t.Deployment)
 	framework.ExpectNoError(err)
 	for _, pod := range pods.Items {
-		t.PodNames = append(t.PodNames, pod.Name)
+		t.Pods = append(t.Pods, PodDetails{Name: pod.Name})
```
Comment (edreed) [inline test/types/resource_types.go:482]: If you're only saving the pod name, why this change to wrap it in `PodDetails`?

**Review by edreed — APPROVED** [review]
